### DISCUSSION TITLE
How to structure mocks that have value type dependencies (i.e., One-to-Many relationships)?

### DISCUSSION THREAD

**Discussion by evankline** [pointfreeco/swift-composable-architecture#2757] [Q&A] [answered]
Hi, I'm new to Swift and TCA and I had planned to use structs as models as I've seen in the examples provided. I'm attempting to preview some of the features and have modeled relationships either with the entire model type or the model type's id value. However, I'm getting a Circular Reference error attempting to model the relationship with mocks for Previews. 

For example, I have the concept of an `Account` and `User` that have a One-to-One relationship:

```
public struct User: Equatable, Identifiable, Codable, Sendable {
  public typealias Id = Tagged<Self, UUID>
  
  public var id: Id
  public var name: String
  public var photoURL: URL?
  public var accountId: Account.Id

  public init(id: Id, name: String = "", photoURL: URL? = nil, accountId: Account.Id) {
    self.id = id
    self.name = name
    self.photoURL = photoURL
    self.accountId = accountId
  }
}

extension User {
  public static let mockOne = User(
    id: User.Id(rawValue: UUID.userOneId),
    accountId: Account.mockOne.id,
    name: "Fred Hunt",
    photoURL: URL(string: "https://randomuser.me/api/portraits/men/25.jpg"),
    companyName: "Fred Hunt, LLC."
  )
}

public struct Account: Equatable, Identifiable, Codable, Sendable {
  public typealias Id = Tagged<Self, UUID>
  
  public var id: Id
  public var email: String
  public var createdAt: Date = .now
  public var updatedAt: Date = .now
  public var lastSignInDate: Date = .now
  public var userId: User.Id

  public init(id: Id, email: String = "", userId = User.Id) {
    self.id = id
    self.email = email
    self.userId = userId
  }
}

extension Account {
  public static let mockOne = Account(
    id: Account.ID(),
    email: "fhunt@example.com",
    userId: User.mockOne.id,
  )
}
```

I'd love some advice on how to model data using TCA and any advice on how to resolve the Circular Reference issue. Thank you!

**Comment by mbrandonw** [comment, marked as answer]
Hi @evankline, I don't think this has anything to do with TCA. You do have a circular reference in your code. `User.mockOne` is using `Account.mockOne`, but also `Account.mockOne` is using `User.mockOne`. That's can't possibly resolve, and so Swift just won't compile it.

If this is just mock data then you can just hard code the IDs however you want, or also I believe if you used `UUID.userOneId` in `Account.mockOne` instead of calling out to `User.mockOne.id` then it would also compile.

**Reply by evankline** [reply]
Hi @mbrandonw, sorry for not clarifying that this isn't a TCA question. I've been learning to use mocks via the Point-Free videos and was more so seeking advice on:

1. What the idiomatic way to structure relationships with using value types rather than reference types (which I know is also not a TCA-specific question)

2. How to do so using Tagged ID values, because when I attempt to use UUID.userOneId I receive the error message:

```
Cannot convert value of type 'UUID' to expected argument type 'User.Id' (aka 'Tagged<User, UUID>')
```

Where `UUID.userOneId` is the following:

```
public extension UUID {
  static let userOneId = Self(uuidString: "foosball-foos-ball-foos-ballfoosball")!
}
```

**Reply by evankline** [reply]
I think I see what you meant now by using `userOneId` and have resolved the issue. Thank you!

**Reply by mbrandonw** [reply]
> * What the idiomatic way to structure relationships with using value types rather than reference types (which I know is also not a TCA-specific question)

There really is no way to do this with value types because value types get copied when passed around. However, we are announcing a new beta period for some tools that deal with this exact issue in the next week or two, so if you can wait till then things will be a lot clearer. The tools will allow you to mostly use value types and insert small references where necessary to express sharing of state.

**Reply by evankline** [reply]
Awesome, looking forward to it... and thanks so much for all the great videos, they're well worth the subscription cost!

**Reply by evankline** [reply]
Awesome, looking forward to it... and thanks so much for all the great videos, they're well worth the subscription cost!
